### PR TITLE
🌱 machine: prevent error spamming for NodeOutdatedTaint if objects are not found

### PR DESCRIPTION
**Pull request by chrischdi**
**What this PR does / why we need it**:

This PR reduces spammy output especially during background deletion of objects like a whole cluster.

Example from https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api/11146/pull-cluster-api-e2e-blocking-main/1832066840088547328/artifacts/clusters/bootstrap/logs/capi-system/capi-controller-manager/capi-controller-manager-5f5f6bcf4c-4zpjc/manager.log :

Example: owner reference is gone:
```
{"ts":1725634615581.657,"caller":"controller/controller.go:316","msg":"Reconciler error","controller":"machine","controllerGroup":"cluster.x-k8s.io","controllerKind":"Machine","Machine":{"name":"quick-start-0qi0bi-md-0-kskml-96tqh-jg4tt","namespace":"quick-start-ev3dh5"},"namespace":"quick-start-ev3dh5","name":"quick-start-0qi0bi-md-0-kskml-96tqh-jg4tt","reconcileID":"5a5d16c2-944d-4bbb-ae38-670497f71998","err":"failed to reconcile Node quick-start-0qi0bi-md-0-kskml-96tqh-jg4tt: failed to check if Node quick-start-0qi0bi-md-0-kskml-96tqh-jg4tt is outdated: failed to find MachineSet owner reference for Machine quick-start-ev3dh5/quick-start-0qi0bi-md-0-kskml-96tqh-jg4tt","errCauses":[{"error":"failed to reconcile Node quick-start-0qi0bi-md-0-kskml-96tqh-jg4tt: failed to check if Node quick-start-0qi0bi-md-0-kskml-96tqh-jg4tt is outdated: failed to find MachineSet owner reference for Machine quick-start-ev3dh5/quick-start-0qi0bi-md-0-kskml-96tqh-jg4tt","errorVerbose":"failed to find MachineSet owner reference for Machine quick-start-ev3dh5/quick-start-0qi0bi-md-0-kskml-96tqh-jg4tt\nsigs.k8s.io/cluster-api/internal/controllers/machine.getOwnerMachineSetObjectKey\n\tsigs.k8s.io/cluster-api/internal/controllers/machine/machine_controller_noderef.go:363\nsigs.k8s.io/cluster-api/internal/controllers/machine.shouldNodeHaveOutdatedTaint\n\tsigs.k8s.io/cluster-api/internal/controllers/machine/machine_controller_noderef.go:323\nsigs.k8s.io/cluster-api/internal/controllers/machine.(*Reconciler).patchNode\n\tsigs.k8s.io/cluster-api/internal/controllers/machine/machine_controller_noderef.go:299\nsigs.k8s.io/cluster-api/internal/controllers/machine.(*Reconciler).reconcileNode\n\tsigs.k8s.io/cluster-api/internal/controllers/machine/machine_controller_noderef.go:139\nsigs.k8s.io/cluster-api/internal/controllers/machine.(*Reconciler).reconcile\n\tsigs.k8s.io/cluster-api/internal/controllers/machine/machine_controller.go:316\nsigs.k8s.io/cluster-api/internal/controllers/machine.(*Reconciler).Reconcile\n\tsigs.k8s.io/cluster-api/internal/controllers/machine/machine_controller.go:240\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).Reconcile\n\tsigs.k8s.io/controller-runtime@v0.19.0/pkg/internal/controller/controller.go:116\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).reconcileHandler\n\tsigs.k8s.io/controller-runtime@v0.19.0/pkg/internal/controller/controller.go:303\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).processNextWorkItem\n\tsigs.k8s.io/controller-runtime@v0.19.0/pkg/internal/controller/controller.go:263\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).Start.func2.2\n\tsigs.k8s.io/controller-runtime@v0.19.0/pkg/internal/controller/controller.go:224\nruntime.goexit\n\truntime/asm_amd64.s:1695\nfailed to check if Node quick-start-0qi0bi-md-0-kskml-96tqh-jg4tt is outdated\nsigs.k8s.io/cluster-api/internal/controllers/machine.(*Reconciler).patchNode\n\tsigs.k8s.io/cluster-api/internal/controllers/machine/machine_controller_noderef.go:301\nsigs.k8s.io/cluster-api/internal/controllers/machine.(*Reconciler).reconcileNode\n\tsigs.k8s.io/cluster-api/internal/controllers/machine/machine_controller_noderef.go:139\nsigs.k8s.io/cluster-api/internal/controllers/machine.(*Reconciler).reconcile\n\tsigs.k8s.io/cluster-api/internal/controllers/machine/machine_controller.go:316\nsigs.k8s.io/cluster-api/internal/controllers/machine.(*Reconciler).Reconcile\n\tsigs.k8s.io/cluster-api/internal/controllers/machine/machine_controller.go:240\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).Reconcile\n\tsigs.k8s.io/controller-runtime@v0.19.0/pkg/internal/controller/controller.go:116\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).reconcileHandler\n\tsigs.k8s.io/controller-runtime@v0.19.0/pkg/internal/controller/controller.go:303\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).processNextWorkItem\n\tsigs.k8s.io/controller-runtime@v0.19.0/pkg/internal/controller/controller.go:263\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).Start.func2.2\n\tsigs.k8s.io/controller-runtime@v0.19.0/pkg/internal/controller/controller.go:224\nruntime.goexit\n\truntime/asm_amd64.s:1695\nfailed to reconcile Node quick-start-0qi0bi-md-0-kskml-96tqh-jg4tt\nsigs.k8s.io/cluster-api/internal/controllers/machine.(*Reconciler).reconcileNode\n\tsigs.k8s.io/cluster-api/internal/controllers/machine/machine_controller_noderef.go:140\nsigs.k8s.io/cluster-api/internal/controllers/machine.(*Reconciler).reconcile\n\tsigs.k8s.io/cluster-api/internal/controllers/machine/machine_controller.go:316\nsigs.k8s.io/cluster-api/internal/controllers/machine.(*Reconciler).Reconcile\n\tsigs.k8s.io/cluster-api/internal/controllers/machine/machine_controller.go:240\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).Reconcile\n\tsigs.k8s.io/controller-runtime@v0.19.0/pkg/internal/controller/controller.go:116\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).reconcileHandler\n\tsigs.k8s.io/controller-runtime@v0.19.0/pkg/internal/controller/controller.go:303\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).processNextWorkItem\n\tsigs.k8s.io/controller-runtime@v0.19.0/pkg/internal/controller/controller.go:263\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).Start.func2.2\n\tsigs.k8s.io/controller-runtime@v0.19.0/pkg/internal/controller/controller.go:224\nruntime.goexit\n\truntime/asm_amd64.s:1695"}]}
```

Example MachineDeployment gone:

```
{"ts":1725634762385.2092,"caller":"controller/controller.go:316","msg":"Reconciler error","controller":"machine","controllerGroup":"cluster.x-k8s.io","controllerKind":"Machine","Machine":{"name":"quick-start-0qi0bi-md-0-kskml-96tqh-jg4tt","namespace":"quick-start-ev3dh5"},"namespace":"quick-start-ev3dh5","name":"quick-start-0qi0bi-md-0-kskml-96tqh-jg4tt","reconcileID":"42367c14-daab-42f7-9a3f-df199c620093","err":"failed to reconcile Node quick-start-0qi0bi-md-0-kskml-96tqh-jg4tt: failed to check if Node quick-start-0qi0bi-md-0-kskml-96tqh-jg4tt is outdated: MachineDeployment.cluster.x-k8s.io \"quick-start-0qi0bi-md-0-kskml\" not found","errCauses":[{"error":"failed to reconcile Node quick-start-0qi0bi-md-0-kskml-96tqh-jg4tt: failed to check if Node quick-start-0qi0bi-md-0-kskml-96tqh-jg4tt is outdated: MachineDeployment.cluster.x-k8s.io \"quick-start-0qi0bi-md-0-kskml\" not found","errorVerbose":"MachineDeployment.cluster.x-k8s.io \"quick-start-0qi0bi-md-0-kskml\" not found\nfailed to check if Node quick-start-0qi0bi-md-0-kskml-96tqh-jg4tt is outdated\nsigs.k8s.io/cluster-api/internal/controllers/machine.(*Reconciler).patchNode\n\tsigs.k8s.io/cluster-api/internal/controllers/machine/machine_controller_noderef.go:301\nsigs.k8s.io/cluster-api/internal/controllers/machine.(*Reconciler).reconcileNode\n\tsigs.k8s.io/cluster-api/internal/controllers/machine/machine_controller_noderef.go:139\nsigs.k8s.io/cluster-api/internal/controllers/machine.(*Reconciler).reconcile\n\tsigs.k8s.io/cluster-api/internal/controllers/machine/machine_controller.go:316\nsigs.k8s.io/cluster-api/internal/controllers/machine.(*Reconciler).Reconcile\n\tsigs.k8s.io/cluster-api/internal/controllers/machine/machine_controller.go:240\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).Reconcile\n\tsigs.k8s.io/controller-runtime@v0.19.0/pkg/internal/controller/controller.go:116\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).reconcileHandler\n\tsigs.k8s.io/controller-runtime@v0.19.0/pkg/internal/controller/controller.go:303\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).processNextWorkItem\n\tsigs.k8s.io/controller-runtime@v0.19.0/pkg/internal/controller/controller.go:263\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).Start.func2.2\n\tsigs.k8s.io/controller-runtime@v0.19.0/pkg/internal/controller/controller.go:224\nruntime.goexit\n\truntime/asm_amd64.s:1695\nfailed to reconcile Node quick-start-0qi0bi-md-0-kskml-96tqh-jg4tt\nsigs.k8s.io/cluster-api/internal/controllers/machine.(*Reconciler).reconcileNode\n\tsigs.k8s.io/cluster-api/internal/controllers/machine/machine_controller_noderef.go:140\nsigs.k8s.io/cluster-api/internal/controllers/machine.(*Reconciler).reconcile\n\tsigs.k8s.io/cluster-api/internal/controllers/machine/machine_controller.go:316\nsigs.k8s.io/cluster-api/internal/controllers/machine.(*Reconciler).Reconcile\n\tsigs.k8s.io/cluster-api/internal/controllers/machine/machine_controller.go:240\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).Reconcile\n\tsigs.k8s.io/controller-runtime@v0.19.0/pkg/internal/controller/controller.go:116\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).reconcileHandler\n\tsigs.k8s.io/controller-runtime@v0.19.0/pkg/internal/controller/controller.go:303\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).processNextWorkItem\n\tsigs.k8s.io/controller-runtime@v0.19.0/pkg/internal/controller/controller.go:263\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).Start.func2.2\n\tsigs.k8s.io/controller-runtime@v0.19.0/pkg/internal/controller/controller.go:224\nruntime.goexit\n\truntime/asm_amd64.s:1695"}]}
```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

/area machine

### CODE REVIEW

**Comment by sbueringer** [comment]
The PR title is a bit misleading :)

**Review by fabriziopandini — COMMENTED** [review]
Nice!
/lgtm

File: internal/controllers/machine/machine_controller_noderef.go [hunk 1]
```diff
@@ -296,10 +297,17 @@ func (r *Reconciler) patchNode(ctx context.Context, remoteClient client.Client,
...
-	isOutdated, err := shouldNodeHaveOutdatedTaint(ctx, r.Client, m)
+	isOutdated, notFound, err := shouldNodeHaveOutdatedTaint(ctx, r.Client, m)
 	if err != nil {
 		return errors.Wrapf(err, "failed to check if Node %s is outdated", klog.KRef("", node.Name))
 	}
+
+	// It is not possible to identify if we have to set the NodeOutdatedRevisionTaint if there is
+	// no OwnerReference or the owning MachineSet or MachineDeployment was not found.
+	// Note: This could happen e.g. during background deletion of objects.
+	if notFound {
```
Comment (sbueringer) [inline internal/controllers/machine/machine_controller_noderef.go:308]: /hold

I think we're now skipping patches if hasAnnotationChanges or hasLabelChanges is true

I think we should only skip the if/else with isOutdated


(Please also check the test coverage to cover these cases correctly)

Comment (chrischdi) [reply internal/controllers/machine/machine_controller_noderef.go]: added the test-case and adjusted the code+comment

Comment (sbueringer) [reply internal/controllers/machine/machine_controller_noderef.go]: Yup good point. But yeah let's improve that as well :D

File: internal/controllers/machine/machine_controller_noderef.go [hunk 2]
```diff
@@ -296,69 +297,84 @@ func (r *Reconciler) patchNode(ctx context.Context, remoteClient client.Client,
...
+	// found all relevant objects.
+	// Example: when the MachineDeployment or Machineset can't be found due to a background deletion of objects.
+	if !notFound {
+		if isOutdated {
+			hasTaintChanges = taints.EnsureNodeTaint(newNode, clusterv1.NodeOutdatedRevisionTaint) || hasTaintChanges
+		} else {
+			hasTaintChanges = taints.RemoveNodeTaint(newNode, clusterv1.NodeOutdatedRevisionTaint) || hasTaintChanges
+		}
+
+		if !hasAnnotationChanges && !hasLabelChanges && !hasTaintChanges {
```
[the remark this answers is not in the document]
Comment (chrischdi) [reply internal/controllers/machine/machine_controller_noderef.go]: the new test-case was just wrong...

File: internal/controllers/machine/machine_controller_noderef_test.go [hunk 3]
```diff
@@ -770,6 +773,63 @@ func TestPatchNode(t *testing.T) {
...
+						Kind:       "MachineSet",
+						Name:       "test-ms-missing",
+						APIVersion: clusterv1.GroupVersion.String(),
+						UID:        "uid",
+					}},
+				},
+				Spec: newFakeMachineSpec(metav1.NamespaceDefault, clusterName),
+			},
+			ms: nil,
+			md: &clusterv1.MachineDeployment{
```
Comment (sbueringer) [inline internal/controllers/machine/machine_controller_noderef_test.go:817]: Maybe set md also to nil. As that's closer to the real-life case

**Comment by sbueringer** [comment]
Thank you!!

/lgtm
/test pull-cluster-api-e2e-main

**Pull request merged**